### DISCUSSION TITLE
Calculation of partial excess energies

### DISCUSSION THREAD

**Discussion by nils-ellendt** [pycalphad/pycalphad#509] [Q&A] [answered]
Dear all,

I am a new user to pycalphad. I have followed the tutorials and I want to implement the butler equation to estimate the surface tension of alloys. I have followed the viscosity example, which is a great help!

I am stuck with the calculation of the partial excess energies. I am using the Sn-Bi system for development because I have a working reference code here which uses the RK directly instead of a tdb.

In my model, I have calculated excess energy with:

G_E=self.excess_mixing_energy(dbe);

which gives me
1.0*(LIQUID0BI*LIQUID0SN*(490.0 + 0.97*T) + (LIQUID0BI - LIQUID0SN)*LIQUID0BI*LIQUID0SN*(-30.0 - 0.235*T))/(LIQUID0BI + LIQUID0SN)

which is fine. How can I use G_E.diff() to calculate the partial excess energies? In the viscosity example

excess_entropy = -excess_energy.diff(v.T)

was used, however, I cannot find entries in V which contain LIQUID0BI or LIQUID0SN. Or is there a smarter way to do that?

thank you for your help!

Nils

**Comment by richardotis** [comment, marked as answer]
Hi Nils, I'm not sure if I fully understand your question, but if you want to calculate partial quantities with respect to the temperature and the site fractions, the following should work:
```python
G_E.diff(v.T)
G_E.diff(v.Y('LIQUID', 0, 'BI'))
G_E.diff(v.Y('LIQUID', 0, 'SN'))
```

**Reply by nils-ellendt** [reply]
Oh, I should have tried this notation. Then I can also grab the elements from self.pure_elements if I change the system, right?

Thank you a lot!

**Reply by richardotis** [reply]
You can use `self.pure_elements`, or `self.nonvacant_elements` if you want vacancies to be excluded (not an issue for liquids usually, unless you're using a model with complex ordering). You can also use `self.site_fractions`, which simply provides a list of all the site fraction variables (the `v.Y(...)`) already created for you, in sublattice-followed-by-alphabetical order.

**Comment by nils-ellendt** [comment]
Some time since I posted here... I have been able to implement the butler equation, a different viscosity model and a model for liquid density. We used it to calculate the effect of Iron contamination of the fluid properties of Al-Si alloys (https://onlinelibrary.wiley.com/doi/10.1002/adem.202401541)

I will clean up the code and then make it available on github
